src/edu/ucsb/nceas/metacat/restservice/multipart/StreamingMultipartRequestResolver.java

```diff
@@ -0,0 +1,277 @@
+                        writeStreamToFile(newFile, stream);
+                        Checksum checksum = null;//we don't have a checksum, so set it null.
+                        CheckedFile checkedFile = new CheckedFile(newFile.getCanonicalPath(), checksum);
+                        mpFiles.put(name, checkedFile);
+                    }
+                } else {
+                    File newFile = generateTmpFile("other");
+                    writeStreamToFile(newFile, stream);
+                    mpFiles.put(name, newFile);
+                }
+            }
+            if(stream != null) {
+                try {
+                    stream.close();
+                } catch (Exception e) {
+                    log.warn("Couldn't close the stream since" + e.getMessage());
+                }
+            }
+        }
+        return multipartRequest;
+    }
+    
+    /**
+     * Create a temporary new file
+     * @return
+     * @throws IOException
+     */
+    private File generateTmpFile(String prefix) throws IOException {
+        String newPrefix = prefix + "-" + System.currentTimeMillis();
+        String suffix =  null;
+        File newFile = null;
+        try {
+            newFile = File.createTempFile(newPrefix, suffix, tempDir);
+        } catch (Exception e) {
+            //try again if the first time fails
+            newFile = File.createTempFile(newPrefix, suffix, tempDir);
+        }
+        log.debug("StreamingMultiplePartRequestResolver.generateTmepFile - the new file  is " + newFile.getCanonicalPath());
+        return newFile;
+    }
+    
+    /**
+     * Write the input stream into the given fileName and directory while calculate the checksum.
+     * @param file  the file into which the stream will be written. It should exists already.
+     * @param dataStream  the source stream
+     * @param checksumAlgorithm  the algorithm will be used for calculating the checksum
+     * @param pid  the pid of the object (only used for debug information)
+     * @return  a CheckedFile object ( a File object with advertised checksum)
+     * @throws NoSuchAlgorithmException
+     * @throws FileNotFoundException
+     * @throws IOException
+     */
+    public static CheckedFile writeStreamToCheckedFile(File file, InputStream dataStream, String checksumAlgorithm, String pid) 
+        throws NoSuchAlgorithmException, FileNotFoundException, IOException {
+        Checksum checksum = null;
+        log.debug("StreamingMultipartRequestResolver.writeStreamToCheckedFile - filename for writting is: " + file.getAbsolutePath() + " for the pid " + pid + " by the algorithm " + checksumAlgorithm);
+        MessageDigest md = MessageDigest.getInstance(checksumAlgorithm);
+        // write data stream to desired file
+        DigestOutputStream os = new DigestOutputStream( new FileOutputStream(file), md);
+        long length = IOUtils.copyLarge(dataStream, os);
+        os.flush();
+        os.close();
```

Comment:
@taojing2002 , I feel like we should be closing these output streams (and others throughout this new code), in `finally` blocks so it isn't left open during exceptions.  This is a potential memory leak I think. Thoughts?